inspire-atom/src/main/java/org/fao/geonet/inspireatom/util/InspireAtomUtil.java

```diff
@@ -305,4 +370,214 @@ public static String retrieveDatasetUuidFromIdentifier(ServiceContext context,
+		}
+	}
+
+	public static Element extractDatasetEntryInfo(final String schema, final Element md,
+			final DataManager dm) throws Exception {
+        java.nio.file.Path styleSheet = dm.getSchemaDir(schema).resolve(EXTRACT_DATASET_ENTRY_INFO);
+        Element datasetEl = Xml.transform(md, styleSheet);
+		md.detach();
+		return datasetEl;
+	}
+
+	public static boolean isAtomDownloadServiceMetadata(
+			DataManager dm, String schema ,
+			Element md, String atomProtocol) throws Exception {
+        java.nio.file.Path styleSheet = dm.getSchemaDir(schema).resolve(EXTRACT_SERVICE_TYPE);
+		Element service = Xml.transform(md, styleSheet);
+		String serviceType = service.getChildText("serviceType").trim();
+		String serviceTypeVersion = service.getChildText("serviceTypeVersion").trim();
+		return atomProtocol.equalsIgnoreCase(serviceType)
+				&& ATOM_SERVICE_TYPE_VERSION.equalsIgnoreCase(serviceTypeVersion);
+	}
+
+	public static List<String> extractRelatedDatasetFileIdentifiers(
+			final String schema, final Element md,
+			final DataManager dm) throws Exception {
+        java.nio.file.Path styleSheet = dm.getSchemaDir(schema).resolve(EXTRACT_DATASET_FILEIDENTIFIERS);
+		@SuppressWarnings("unchecked")
+		List<Element> fileIdentifiersEl = Xml.transform(md, styleSheet)
+				.getChildren();
+		List<String> fileIdentifiers = new ArrayList<String>();
+
+		// --- needed to detach md from the document
+		md.detach();
+
+		for (Element fileIdentifierEl : fileIdentifiersEl) {
+			String fileIdentifier = fileIdentifierEl.getText();
+
+			if (!StringUtils.isEmpty(fileIdentifier))
+				fileIdentifiers.add(fileIdentifier);
+		}
+
+		return fileIdentifiers;
+	}
+	
+	public static Element getDatasetFeed(DataManager dm, ServiceContext context, String metadataId, String fileIdentifier, String datasetIdCode, String datasetIdNs, String crs, String searchTerms)
+			throws Exception {
+		if (!StringUtils.isBlank(fileIdentifier)) {
+			Element md = dm.getMetadata(metadataId);
+            String schema = dm.getMetadataSchema(metadataId);
+			if (!InspireAtomUtil.isDatasetMetadata(dm, schema, md)) {
+				throw new Exception("No dataset metadata found with uuid:"
+						+ fileIdentifier);
+			}
+			boolean bCodeSpaceValueIsEqual = false;
+			if (!StringUtils.isBlank(datasetIdNs)) {
+				List<Namespace> nss = new ArrayList<Namespace>();
+//				nss.addAll(md.getAdditionalNamespaces());
+//				nss.add(md.getNamespace());
+				nss.add(Namespace.getNamespace("gmd","http://www.isotc211.org/2005/gmd"));
+				nss.add(Namespace.getNamespace("gco","http://www.isotc211.org/2005/gco"));
+				Object o = Xml
+						.selectSingle(
+								md,
+								"gmd:identificationInfo/gmd:MD_DataIdentification/gmd:citation/gmd:CI_Citation/gmd:identifier/gmd:MD_Identifier/gmd:codeSpace/gco:CharacterString|gmd:identificationInfo/gmd:MD_DataIdentification/gmd:citation/gmd:CI_Citation/gmd:identifier/gmd:RS_Identifier/gmd:codeSpace/gco:CharacterString",
```

Comment:
This won't work when having non ISO19139 records (even if ISO19139 compatible eg. ISO19115-3). This part should be pluggable from the Schema and should not use hardcoded Namespaces and XPaths  eg. https://github.com/geonetwork/core-geonetwork/pull/1280